src/json-patch-duplex.ts

```diff
@@ -524,95 +513,263 @@ module jsonpatch {
+        return returnValue;
+      }
+      else if (operation.op === 'move' || operation.op === 'copy') { // it's a move or copy to root
+        returnValue.newDocument = getValueByPointer(document, operation.from); // get the value by json-pointer in `from` field
+        if (operation.op === 'move') { // report removed item
+          returnValue.result = document;
+        }
+        return returnValue;
+      } else { // it's a remove or test on root
+        if (operation.op === 'test') {
+          returnValue.result = _equals(document, operation.value);
+          if(returnValue.result == false) {            
+            throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
+          }
+          returnValue.newDocument = document;
+          return returnValue;
+        } else if (operation.op === 'remove') { // a remove on root
+          returnValue.result = document;
+          returnValue.newDocument = null;
+          return returnValue;
+        } else { /* bad operation */
+          if (validateOperation) {
+            throw new JsonPatchError('Operation `op` property is not one of operations defined in RFC-6902', 'OPERATION_OP_INVALID', 0, operation, document);
+          } else {
+            returnValue.newDocument = document;
+            return returnValue;
+          }
+        }
+      }
+    } /* END ROOT OPERATIONS */
+    else {
+      if (!mutateDocument) {
+        document = deepClone(document);
+      }
+      const path = operation.path || "";
+      const keys = path.split('/');
+      let obj = document;
+      let t = 1; //skip empty element - http://jsperf.com/to-shift-or-not-to-shift
+      let len = keys.length;
+      let existingPathFragment = undefined;
+      let key: string | number;
       while (true) {
         key = keys[t];
 
-        if (validate) {
+        if (validateOperation) {
           if (existingPathFragment === undefined) {
             if (obj[key] === undefined) {
               existingPathFragment = keys.slice(0, t).join('/');
             }
             else if (t == len - 1) {
-              existingPathFragment = patch.path;
+              existingPathFragment = operation.path;
             }
             if (existingPathFragment !== undefined) {
-              this.validator(patch, p - 1, tree, existingPathFragment);
+              if(typeof validateOperation == 'function') {
+                validateOperation(operation, 0, document, existingPathFragment);
+              }
+              else {      
+                validator(operation, 0, document, existingPathFragment);
+              }
             }
           }
         }
-
         t++;
-        if (key === undefined) { //is root
-          if (t >= len) {
-            results.push(rootOps[patch.op].call(patch, obj, key, tree)); // Apply patch
-            break;
-          }
-        }
         if (_isArray(obj)) {
+          let length = obj.length;
```

Comment:
Previously we were avoiding checking `obj.length` for significant (in our cases bigger) amount of cases: `key!='-' && ! validate`, and doing it twice for narrow case `key='-' && validate`.

Now, it always checks once.

Assuming, the user focused on performance sets `validate` to `false`, we have worsen his/hers performance.